### PR TITLE
Reset Username API

### PR DESCRIPTION
**Pull request by movshov**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

Allows to change an Authority's Username. 

* **What is the current behavior?** (You can also link to an open issue here)

Username is Static

* **What is the new behavior (if this is a feature change)?**

Username will be able to get updated. 

* **Other information**:

### CODE REVIEW

File: api/accounts/account-update-username.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+export default handler(async (req, userId): Promise<string> => {
+    errors.requireParams(req.body, ['username','newusername'],);
+    const username = req.body.username;
+    const newname = req.body.newusername;
+    return withConnection(async (db: Client) => {
+        const matchingAccount = await db.query('SELECT username FROM authority WHERE username = $1', [username]);
+        if (matchingAccount.rows.length !== 1) {
+            throw new errors.BadParameter('No such account with username ' + username);
+        }
+        await db.query('UPDATE authority SET username = $1 WHERE username = $2', [newname, username]);
```
Comment (ElijahRW) [inline api/accounts/account-update-username.ts]: You can take advantage of that PG will return the rows of the UPDATE query. So you only need one query to the DB: Try this instead:
```suggestion
        const updateResults = await db.query('UPDATE authority SET username = $1 WHERE username = $2', [newname, username]);
        if (updateResults.rowCount !== 1) {
            throw new errors.BadParameter('No such account with username ' + username);
        }
```

File: api/app.ts [hunk 2]
```diff
@@ -30,6 +30,9 @@ app.post('/changePassword', changePasswordEndpoint);
 import accountCreate from './accounts/create';
 app.post('/accounts/create', accountCreate);
 
+import resetUsername from './accounts/account-update-username';
+app.post('/accounts/resetUsername', resetUsername);
```
Comment (ElijahRW) [inline api/app.ts]: endpoint should hit the endpoint `accounts/changeUserName`.
```suggestion
app.post('/accounts/changeUsername', changeUsername);
```

**Review by ElijahRW — CHANGES_REQUESTED** [review]
This file should be renamed to `changeUsername.ts` to follow the rest of our naming convention in the API folder.

File: api/accounts/account-update-username.ts (hunk 1 above)
Comment (movshov) [reply api/accounts/account-update-username.ts]: okay will do.

File: api/app.ts (hunk 2 above)
Comment (movshov) [reply api/app.ts]: fixed.

File: api/accounts/changeUsername.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    const username = req.body.username;
+    const newname = req.body.newusername;
+    return withConnection(async (db: Client) => {
+        const matchingAccount = await db.query('SELECT username FROM authority WHERE username = $1', [username]);
+        if (matchingAccount.rows.length !== 1) {
+            throw new errors.BadParameter('No such account with username ' + username);
+        }
+        const updateResults = await db.query('UPDATE authority SET username = $1 WHERE username = $2', [newname, username]);
+        if (updateResults.rowCount !== 1){
+            throw new errors.BadParameter('No such account with username ' + username);
```
Comment (gstone271) [inline api/accounts/changeUsername.ts:16]: This error is misleading. Instead throw an APIError that indicates that this should never happen ("Somehow failed to...")

Comment (movshov) [reply api/accounts/changeUsername.ts:16]: Not quite sure what you mean by APIError?

Comment (ElijahRW) [reply api/accounts/changeUsername.ts:16]: (I clarified with Bar: I think we should remove the `matchingAccount` SQL call and check and just use this check to throw the bad parameter error.

File: api/app.ts [hunk 4]
```diff
@@ -30,6 +30,9 @@ app.post('/changePassword', changePasswordEndpoint);
 import accountCreate from './accounts/create';
 app.post('/accounts/create', accountCreate);
 
+import changeUsername from './accounts/changeUsername';
+app.post('/accounts/changeUsername', changeUsername);
```
Comment (Liriaene) [inline api/app.ts:34]: I think this might need to be a PUT call since we aren't creating new rows, just updating.

Comment (ElijahRW) [reply api/app.ts:34]: I think we should leave it as a post for now. If we wish to change the endpoint, we can at a future time.

File: api/accounts/changeUsername.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import { Client } from 'pg';
+
+// Expects 'username'
+// Returns the randomly generated password
+export default handler(async (req, userId): Promise<string> => {
+    errors.requireParams(req.body, ['username','newusername'],);
+    const username = req.body.username;
+    const newname = req.body.newusername;
+    return withConnection(async (db: Client) => {
+        const matchingAccount = await db.query('SELECT username FROM authority WHERE username = $1', [username]);
```
Comment (ElijahRW) [inline api/accounts/changeUsername.ts]: To get this PR approved: please resolve this redundant error check.

**Review by ElijahRW — CHANGES_REQUESTED** [review]

File: api/app.ts (hunk 4 above)
Comment (gstone271) [reply api/app.ts:34]: This is not a PUT call; it does not create new data and is not idempotent.

**Review by ElijahRW — APPROVED** [review]

**Pull request merged**